examples/viewer.js

```diff
@@ -223,18 +473,122 @@ async function startViewer(localView, remoteView, formValues, onStatsReport, onR
             iceTransportPolicy: formValues.forceTURN ? 'relay' : 'all',
         };
         viewer.peerConnection = new RTCPeerConnection(configuration);
+
+        viewer.peerConnection.onicegatheringstatechange = (event) => {
+            if (viewer.peerConnection.iceGatheringState === 'gathering') {
+                metrics.viewer.iceGathering.startTime = Date.now();
+            } else if (viewer.peerConnection.iceGatheringState === 'complete') {
+                metrics.viewer.iceGathering.endTime = Date.now();
+            }
+        };
+
+        viewer.peerConnection.onconnectionstatechange = (event) => {
+            if (viewer.peerConnection.connectionState === 'new' || viewer.peerConnection.connectionState === 'connecting') {
+                metrics.viewer.peerConnection.startTime = Date.now();
+            }
+            if (viewer.peerConnection.connectionState === 'connected') {
+                metrics.viewer.peerConnection.endTime = Date.now();
+                metrics.viewer.ttffAfterPc.startTime = metrics.viewer.peerConnection.endTime;
+            }
+        };
+
+        viewer.peerConnection.oniceconnectionstatechange = (event) => {
+            if (viewer.peerConnection.iceConnectionState === 'connected') {
+                viewer.peerConnection.getStats().then(stats => {
+                    stats.forEach(report => {
+                        if (report.type === 'candidate-pair') {
+                            activeCandidatePair = report;
+                        }
+                    });
+                });
+            }
+        };
+
         if (formValues.openDataChannel) {
             const dataChannelObj = viewer.peerConnection.createDataChannel('kvsDataChannel');
             viewer.dataChannel = dataChannelObj;
-            dataChannelObj.onopen = event => {
-                dataChannelObj.send("Opened data channel by viewer");
+            dataChannelObj.onopen = () => {
+                if (formValues.enableDQPmetrics) {
+                    dataChannelLatencyCalcMessage.t1 = Date.now();
+                    dataChannelObj.send(JSON.stringify(dataChannelLatencyCalcMessage));
+                } else {
+                    dataChannelObj.send("Opened data channel by viewer");
+                }
             };
             // Callback for the data channel created by viewer
-            dataChannelObj.onmessage = onRemoteDataMessage;
+            let onRemoteDataMessageViewer = (message) => {
+                
+                remoteMessage.append(`${message.data}\n\n`);
+                if (formValues.enableDQPmetrics) {
+                    try {
+                        let dataChannelMessage = JSON.parse(message.data);
+                        if (dataChannelMessage.hasOwnProperty('t1')) {
+                            test = dataChannelMessage;
+                            if (dataChannelMessage.t3 == '') {
```

Comment:
===

Reply:
let's include the PR explanation of this "The datachannel first sends a message of the following format with t1 attached, to which the master responds back with..." as a comment?

Reply:
Included in the description already

Reply:
When it's cloned locally, we don't really tend to look at the PR description anymore. mostly comments